daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala

```diff
@@ -0,0 +1,349 @@
+            case (
+                  ValueVariant(recordedTyCon, recordedVariant, recordedValue),
+                  ValueVariant(replayedTyCon, replayedVariant, replayedValue),
+                ) =>
+              tyConIsReplayedBy(recordedTyCon, replayedTyCon) &&
+              recordedVariant == replayedVariant
+              loop((recordedValue, replayedValue) +: rest)
+            case (ValueList(recordedValues), ValueList(replayedValues)) =>
+              loop((recordedValues.iterator zip replayedValues.iterator) ++: rest)
+            case (ValueOptional(recordedValue), ValueOptional(replayedValue)) =>
+              (recordedValue, replayedValue) match {
+                case (Some(recorded), Some(replayed)) => loop((recorded, replayed) +: rest)
+                case (None, None) => loop(rest)
+                case _ => false
+              }
+            case (ValueTextMap(recordedEntries), ValueTextMap(replayedEntries)) =>
+              recordedEntries.length == replayedEntries.length &&
+                keys(recordedEntries.toImmArray).sameElements(keys(replayedEntries.toImmArray)) &&
+                loop(
+                  (values(recordedEntries.toImmArray) zip values(replayedEntries.toImmArray)) ++:
+                    rest
+                )
+            case (ValueGenMap(recordedEntries), ValueGenMap(replayedEntries)) =>
+              recordedEntries.length == replayedEntries.length &&
+                loop(
+                  (keys(recordedEntries) zip keys(replayedEntries)) ++:
+                    (values(recordedEntries) zip values(replayedEntries)) ++:
+                    rest
+                )
+            case _ =>
+              false
+          }
+        case LazyList() =>
+          true
+      }
+    loop(LazyList((recorded, replayed)))
+  }
+
+  /** Whether `other` is the result of reinterpreting this transaction.
+    *
+    * @param recorded : the transaction to be validated.
+    * @param replayed : the transaction resulting from the reinterpretation of
+    *   the root nodes of [[recorded]].
+    * @note This function is asymmetric in order to provide backward compatibility.
+    *      For instance, some field may be undefined in the [[recorded]] transaction
+    *      while present in the [[replayed]] one.
+    */
+  private def isReplayedBy(
+      recorded: VersionedTransaction[Nid, Cid],
+      replayed: VersionedTransaction[Nid, Cid],
+  ): Either[ReplayMismatch[Nid, Cid], Unit] = {
+
+    type Exe = Node.NodeExercises[Nid, Cid]
+
+    @tailrec
+    def loop(
+        nids1: LazyList[Nid],
+        nids2: LazyList[Nid],
+        stack: List[(Nid, Exe, LazyList[Nid], Nid, Exe, LazyList[Nid])] = List.empty,
+    ): Either[ReplayMismatch[Nid, Cid], Unit] =
+      (nids1, nids2) match {
+        case (LazyList.cons(nid1, rest1), LazyList.cons(nid2, rest2)) =>
+          (recorded.nodes(nid1), replayed.nodes(nid2)) match {
+            case (
```

Comment:
this seems pretty fragile, e.g, it’s relatively easy to forget a field (I realize it’s not introduced in this PR). Unfortunately, I also don’t have great suggestion for improving it :disappointed: